### DISCUSSION TITLE
Can py-reml show "...working..." while it is doing long computations?

### DISCUSSION THREAD

**Discussion by CarlKCarlK** [pyscript/pyscript#1414] [Q&A] [answered]
On Thursday at PyData, I'm presenting a [physics simulation](https://github.com/CarlKCarlK/perfect-physics) that uses sympy (computer algebra) and matplotlib. It would be great fun to offer a PyScript version. I have [something basic working with py-repl](https://carlkcarlk.github.io/perfect-physics/pyscript/) but ...

* Is there a way to have it show "...working..." while it is evaluating the user's repl expression? (The evaluation can take 10's of seconds.)
* Can it clear the output \<div\> just before each user evaluation?
* (I'm very open to other suggestions.)

Thanks for any help,
Carl

**Comment by WebReflection** [comment]
AFAIK until the worker based logic is finalized and published, blocking tasks block the main thread so even if the theoretical answer is "*yes, you can*" the practical result would be a frozen page until results are provided.

![image](https://user-images.githubusercontent.com/85749/234521582-b61e9a5f-3c85-4443-8c16-08c05521a41b.png)

Once the worker solution is final, you should be able to:
  * clean the current page before executing the new task
  * show a spinner or anything animated that shouldn't block unless your machine has a single CPU and a single thread 😅
  * drop the spinner when results are provided (i.e. don't append results to the result container, just write into it)

Happy to be corrected if I've said something wrong, but I think it's just a matter of time this list of points would be actually possible.

**Comment by JeffersGlass** [comment]
All of what Andrea said is true - currently the page (main thread) is non-interactable/non-updatable while running synchronous Python code. However if what you're looking for is a simple state change before/after the user's REPL code is run, you might check out an example I wrote for another user to [change the color/opacity of the REPL run button while a user's REPL code is running](https://github.com/pyscript/pyscript/discussions/1418#discussioncomment-5731668), which could be adapted to your purposes.

Note that that sample code relies on using features only present in the `unstable` branch of PyScript (i.e. merged but not released). I would *not* recommend it for any kind of long-term use, but for something to play around with or for a one off, it might be interesting to look at.

**Reply by CarlKCarlK** [reply]
Thanks! Unstable is fine (my code is unstable, too!). If you're available, I have a follow-ups. Please see below.

**Comment by CarlKCarlK** [comment]
Thank you, Jeffers and WebReflection. I've updated it to use Jeffers' suggestion:
https://carlkcarlk.github.io/perfect-physics/pyscript/

A couple of follow ups. 

* Is there any way to pre-import items and pre-define functions outside the REPL? Before I had these in \<pyscript\>, but now they seem to need to be in \<py-repl\>
```
from perfect_physics import World, Circle, Wall
def run_in_place(world, steps):
  world.run_in_place(steps, show=True, show_fun=lambda figure: display(figure, target="replOutput", append=True))

```
* What's a good way to clear the output div between runs? (Complication: The long running job makes multiple calls to display, so it has append=True.)

- Carl

**Reply by JeffersGlass** [reply]
I'm not entirely sure I understand the question about pre-defining functions. Something like this? This currently works for me in unstable:
```html
<py-script>
  def foo():
    return "FOO"
</py-script>
<py-repl>
  print(foo())
</py-repl>
```

----

You want to clear `<div id="replOutput"></div>` every time you run `run_in_place`, yeah? You could do something like:
```python
import js
...
def run_in_place(...):
    js.document.getElementById("replOutput").innerHTML = ''
```

**Reply by CarlKCarlK** [reply]
Jeffers, clearing the div worked. Thanks!
Defining code in \<py-script\> for \<py-reml\> doesn't seem to work with "unstable". Here is an example page:
https://carlkcarlk.github.io/perfect-physics/pyscript/index_temp.html

When I move the code to \<py-script\> it says:
```
Traceback (most recent call last):
  File "/home/pyodide/pyscript/_internal.py", line 104, in run_pyscript
    result = eval_code(code, globals=__main__.__dict__)
  File "/lib/python3.10/_pyodide/_base.py", line 460, in eval_code
    .run(globals, locals)
  File "/lib/python3.10/_pyodide/_base.py", line 306, in run
    coroutine = eval(self.code, globals, locals)
  File "<exec>", line 1, in <module>
NameError: name 'Circle' is not defined
```

**Reply by JeffersGlass** [reply, marked as answer]
Ah, I think perhaps it's an issue in that Plugin I provided. Would you add the following import and adjusted class definition to `replstatus.py`:

```python
...
from pyscript import Plugin

class ReplStatus(Plugin): #Should be a subclass of pyscript's Plugin class
   ...
```

**Reply by CarlKCarlK** [reply]
It works! https://carlkcarlk.github.io/perfect-physics/pyscript/
(For my personal browser, to get the new version of the plugin code, I had to close all instances of the browser.)
- Carl

**Reply by JeffersGlass** [reply]
Excellent! 🥳

On the needing-to-close-tabs thing, depending on your setup, you might need to manually check "Disable Cache" in your browser's developer tools? In Chrome, it's at this location under the network tab... for other browser's it might be worth a google.

![image](https://user-images.githubusercontent.com/1931111/234664518-4a20b833-14ca-4875-8df9-e3cc77ee76cf.png)

And if it's already checked, then I'm baffled. But glad it works!
